examples/advanced/maximal_propagate.py

```diff
@@ -0,0 +1,39 @@
+"""
+Example showing repeated solving for deriving the maximal consequence of set of constraints.
+
+Iteratively finds a new solutions by forcing at least one variable to have a value not seen in any other solution.
+"""
+
+from cpmpy import *
+from cpmpy.transformations.get_variables import get_variables
+
+
+def maximal_propagate(constraints, solvername="ortools"):
+    # For faster propagation, use incremental solver such as pysat, z3 or gurobi.
+    vars = get_variables(constraints)
+    visisted_domain = {var: set() for var in vars}
+
+    solver = SolverLookup.get(solvername)
+    solver += constraints
+
+    while solver.solve():
+        cons = False
+        for var, values in visisted_domain.items():
+            values.add(var.value())
+            cons |= all(var != val for val in values)
+        solver += cons
+    # exhausted all possible domain values
+    return visisted_domain
```

Comment:
I'm not sure if it is better to return assignments or variables and their corresponding domains.

Reply:
Well as it is a propagator, you would want the domains allowed by the constraints you provide, right?

Reply:
Yeah, the restricted domains seems the most flexible choice (and matches nicely with the algorithm).